### PR TITLE
 feat(containers)!: redirect object templates to ObjectPtr storage

### PR DESCRIPTION
**Pull request by junrushao**
Architecture:
- Normalize unqualified Object subclasses through object_ptr_type_t at every
  container storage, conversion, and subsumption boundary.
- Preserve public class templates and C++17 deduction while sharing backing
  storage between shorthand and explicit ObjectPtr spellings.
- Apply the same normalization recursively to Array, List, Map, Dict, Tuple,
  Optional, Variant, and Expected.

Public Interfaces:
- Make Container<TObject> shorthand for Container<ObjectPtr<TObject>> across
  sequence, mapping, tuple, optional, variant, and expected templates.
- Permit zero-copy handle conversion between shorthand and explicit pointer
  spellings when their normalized types subsume one another.

UI/UX:
- None.

Behavioral Changes:
- Container accessors, schemas, checks, and ownership now use the normalized
  pointer type whenever a template argument is an unqualified Object subclass.
- Qualified object types remain unnormalized and unsupported by this shorthand.

Docs:
- Add examples for every redirected container and document storage equivalence,
  zero-copy conversions, deduction behavior, and qualification limits.

Tests:
- Clean Debug tvm_ffi_tests build completed successfully.
- All 444 enabled CTests passed; 2 cycle-detection tests remained disabled.
- Clean editable build and 552 focused container, metadata, and dataclass Python
  tests passed.
- Stub generation was idempotent on the complete tree.
- Scoped pre-commit hooks passed for all files in this boundary.

Untested Edge Cases:
- Rust bindings and the cross-platform CI matrix were not run locally.
- The unchanged original target tree has a known MSVC C2976 failure around
  Array<Variant<int64_t, String>> subsumption; this semantic split does not fix it.
- The unchanged target also has Doxygen warnings for new subsumption/Optional
  specializations and enum APIs; documentation cleanup remains follow-up work.
- Targeted clang-tidy was not rerun during the split; the identical final tree
  had passed the repository's targeted check before history was rewritten.

BREAKING CHANGE: unqualified Object subclasses used as container template
arguments now normalize to ObjectPtr storage. Code depending on the old declared
type behavior must use the explicit intended storage type and rebuild extensions.

### CODE REVIEW

File: python/tvm_ffi/cython/pyclass_type_converter.pxi [hunk 1]
```diff
@@ -36,9 +36,9 @@ cdef object _INT64_MIN = -(1 << 63)
 cdef object _INT64_MAX = (1 << 63) - 1
 cdef int _VALUE_PROTOCOL_MAX_DEPTH = 64
 cdef str _TYPE_ATTR_FFI_CONVERT = "__ffi_convert__"
-cdef str _TYPE_ATTR_ENUM_VALUE_ENTRIES = "__ffi_enum_value_entries__"
+cdef str _TYPE_ATTR_ENUM_STATE = "__ffi_enum__"
 cdef class _TypeConverter
-ctypedef CAny (*_dispatch_fn_t)(_TypeConverter, object, bint*) except *
+ctypedef CAny (*_dispatch_fn_t)(_TypeConverter, object, bint*)
```
[the remark this answers is not in the document]
Comment (junrushao) [reply python/tvm_ffi/cython/pyclass_type_converter.pxi:41]: Thanks for flagging this. This thread is outdated after the PR was narrowed and rebased: `python/tvm_ffi/cython/pyclass_type_converter.pxi` is no longer changed by PR #666 relative to `main`. I’ve therefore left the converter exception specifications unchanged in this PR so the concern can be handled in the change that owns that file.

File: include/tvm/ffi/container/variant.h [hunk 2]
```diff
@@ -167,7 +185,8 @@ class Variant {
    * ObjectRef
    */
   TVM_FFI_INLINE Object* GetObjectPtrForHashEqual() const {
-    constexpr bool all_object_v = (std::is_base_of_v<ObjectRef, V> && ...);
+    constexpr bool all_object_v =
+        (std::is_base_of_v<ObjectRef, details::object_ptr_type_t<V>> && ...);
     static_assert(all_object_v,
                   "All types used in Variant<...> must be derived from ObjectRef "
                   "to enable ObjectPtrHash/ObjectPtrEqual");
```
[the remark this answers is not in the document]
Comment (junrushao) [reply include/tvm/ffi/container/variant.h:202]: Fixed in `bef03d5`. `GetObjectPtrForHashEqual` now checks each normalized alternative’s `TypeTraits<...>::field_static_type_index` against `kTVMFFIObject`, so both `ObjectRef` and `ObjectPtr` alternatives are accepted. I also added `Variant.ObjectSubclassShorthandHashEqual`; the full C++ test suite and targeted clang-tidy pass.

File: include/tvm/ffi/container/variant.h [hunk 3]
```diff
@@ -71,6 +71,14 @@ class Variant {
    */
   Variant(Variant<V...>&& other) noexcept = default;
 
+  /*! \brief Convert from a variant whose normalized alternatives are all contained. */
+  template <typename... U, std::enable_if_t<(variant_contains_v<U> && ...), int> = 0>
+  Variant(const Variant<U...>& other) : data_(Any(other)) {}  // NOLINT(*)
+
+  /*! \brief Move from a variant whose normalized alternatives are all contained. */
+  template <typename... U, std::enable_if_t<(variant_contains_v<U> && ...), int> = 0>
+  Variant(Variant<U...>&& other) : data_(Any(std::move(other))) {}  // NOLINT(*)
```
[the remark this answers is not in the document]
Comment (junrushao) [reply include/tvm/ffi/container/variant.h:84]: Fixed in `bef03d5`. Both converting constructors now require `!std::is_same_v<Variant, Variant<U...>>`, so exact-type copy and move continue to use the default constructors. The full C++ test suite and targeted clang-tidy pass.

File: include/tvm/ffi/container/variant.h [hunk 4]
```diff
@@ -83,6 +91,16 @@ class Variant {
...
 
+  template <typename... U, std::enable_if_t<(variant_contains_v<U> && ...), int> = 0>
+  TVM_FFI_INLINE Variant& operator=(const Variant<U...>& other) {
+    return operator=(Variant(other));
+  }
+
+  template <typename... U, std::enable_if_t<(variant_contains_v<U> && ...), int> = 0>
+  TVM_FFI_INLINE Variant& operator=(Variant<U...>&& other) {
+    return operator=(Variant(std::move(other)));
+  }
```
[the remark this answers is not in the document]
Comment (junrushao) [reply include/tvm/ffi/container/variant.h]: Fixed in `bef03d5`. Both converting assignment operators now require `!std::is_same_v<Variant, Variant<U...>>`, keeping exact-type copy and move on the default assignment operators and avoiding recursive template assignment. The full C++ test suite and targeted clang-tidy pass.
